awx/main/models/credential/__init__.py

```diff
@@ -562,6 +586,11 @@ def inject_credential(self, credential, env, safe_env, args, safe_args, private_
                                  files)
         """
         if not self.injectors:
+            if self.managed_by_tower and credential.kind in dir(builtin_injectors):
```

Comment:
The way I define these reusable functions may be the most controversial thing here - I'm open to other suggestions if you all hate this. If the cred is `managed_by_tower` and there's a function in `awx.main.models.credentials.injectors` with a name that matches `credential.kind`, we call it to perform the injection.

So the "builtin injectors" are defined in code:

https://github.com/ansible/awx/pull/1001/files#diff-24c6c27b21e897c81eb8e906106f1b47

One thought I had was that in the future we could maybe expand on this by giving customers a way to point at these named builtins in their own custom credential types, e.g.,
```
# My Custom Credential Type
injectors = {
    'builtin': 'azure_rm',
    'env': {'AZURE_EXTRA': 'FOO'}
}
```

Reply:
![](https://media1.giphy.com/media/yj5oYHjoIwv28/giphy.gif)

Reply:
💔

Reply:
I'm fine with the function names constructed on the fly. Thanks for pointing out.